include/dlaf/auxiliary/norm/mc.h

```diff
@@ -15,14 +15,39 @@
 #include <dlaf/auxiliary/norm/api.h>
 #include <dlaf/common/range2d.h>
 #include <dlaf/common/vector.h>
+#include <dlaf/communication/kernels/reduce.h>
 #include <dlaf/communication/sync/reduce.h>
 #include <dlaf/lapack/tile.h>
 #include <dlaf/matrix/distribution.h>
+#include <dlaf/sender/transform_mpi.h>
 #include <dlaf/types.h>
 #include <dlaf/util_matrix.h>
 
 namespace dlaf::auxiliary::internal {
 
+template <class T>
+pika::execution::experimental::unique_any_sender<T> reduce_in_place(
+    pika::execution::experimental::unique_any_sender<dlaf::comm::CommunicatorPipelineExclusiveWrapper>
+        pcomm,
+    comm::IndexT_MPI rank, MPI_Op reduce_op, pika::execution::experimental::unique_any_sender<T> value) {
+  namespace ex = pika::execution::experimental;
+
+  return ex::when_all(std::move(value)) |
+         ex::let_value([pcomm = std::move(pcomm), rank, reduce_op](T& local) mutable {
+           using dlaf::comm::internal::transformMPI;
+           return dlaf::internal::whenAllLift(std::move(pcomm)) |
+                  transformMPI([rank, reduce_op, &local](const dlaf::comm::Communicator& comm,
+                                                         MPI_Request* req) mutable {
+                    const bool is_root_rank = comm.rank() == rank;
+                    void* in = is_root_rank ? MPI_IN_PLACE : &local;
+                    void* out = is_root_rank ? &local : nullptr;
+                    DLAF_MPI_CHECK_ERROR(MPI_Ireduce(in, out, 1, dlaf::comm::mpi_datatype<T>::type,
+                                                     reduce_op, rank, comm, req));
+                  }) |
+                  ex::then([&local]() -> T { return local; });
+         });
```

Comment:
I think this looks fine from a lifetime perspective (otherwise too). `let_value` keeps `local` alive until at least when `transformMPI` signals completion (really, until the end of the `then` callable). Since the `then` callable is not going to be called until after the `MPI_Ireduce` is done, we can at that point read what's been written into `local` and return it for further consumption.

Just for completeness I'll note that another possible implementation strategy would be to wrap `value` in a `Tile`, because then one could use `transform_mpi`'s built-in keep-the-reference-alive and remove the outer `let_value` (in fact, `transform_mpi` uses a `let_value` internally for exactly this reason), and then return the `Tile` from the `transform_mpi` callable.  However, we can't actually use after #1125 is merged because pika's `transform_mpi` doesn't allow returning anything except `void` or MPI return codes from the user callable. This is implementable, but was just cut for simplification until we actually need the feature. This second strategy of course also implies dynamic allocation and reference counting for the `Tile`, which may be suboptimal.